# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/style/FontFaceSet.cpp
# Commit: 22f567c00af3
# Full Hash: 22f567c00af321b797590746f181763abb07d83b
# Author: Andrew Osmond <aosmond@mozilla.com>
# Date: 2022-06-29 03:44:30
# Regressor Bug: 1771493
# File Overlap Count: 1
# Description:
#   Bug 1771493 - Part 4. Split FontFaceSetDocumentImpl from FontFaceSetImpl. r=emilio
#   
#   This patch splits document specific functionality from FontFaceSetImpl
#   into a new class FontFaceSetDocumentImpl. This will make it easier to
#   fork FontFaceSetImpl for workers.
# ==============================================================================

diff -r 609a5d93bf08 -r 22f567c00af3 layout/style/FontFaceSet.cpp
--- a/layout/style/FontFaceSet.cpp	Tue Jun 28 15:58:10 2022 +0000
+++ b/layout/style/FontFaceSet.cpp	Tue Jun 28 15:58:11 2022 +0000
@@ -15,6 +15,7 @@
 #include "mozilla/dom/DocumentInlines.h"
 #include "mozilla/dom/Event.h"
 #include "mozilla/dom/FontFaceSetBinding.h"
+#include "mozilla/dom/FontFaceSetDocumentImpl.h"
 #include "mozilla/dom/FontFaceSetIterator.h"
 #include "mozilla/dom/FontFaceSetLoadEvent.h"
 #include "mozilla/dom/FontFaceSetLoadEventBinding.h"
@@ -92,11 +93,8 @@
 NS_INTERFACE_MAP_BEGIN_CYCLE_COLLECTION(FontFaceSet)
 NS_INTERFACE_MAP_END_INHERITING(DOMEventTargetHelper)
 
-FontFaceSet::FontFaceSet(nsPIDOMWindowInner* aWindow, dom::Document* aDocument)
-    : DOMEventTargetHelper(aWindow),
-      mImpl(new FontFaceSetImpl(this, aDocument)) {
-  mImpl->Initialize();
-}
+FontFaceSet::FontFaceSet(nsIGlobalObject* aParent)
+    : DOMEventTargetHelper(aParent) {}
 
 FontFaceSet::~FontFaceSet() {
   // Assert that we don't drop any FontFaceSet objects during a Servo traversal,
@@ -106,6 +104,16 @@
   Destroy();
 }
 
+/* static */ already_AddRefed<FontFaceSet> FontFaceSet::CreateForDocument(
+    dom::Document* aDocument) {
+  RefPtr<FontFaceSet> set = new FontFaceSet(aDocument->GetScopeObject());
+  RefPtr<FontFaceSetDocumentImpl> impl =
+      new FontFaceSetDocumentImpl(set, aDocument);
+  impl->Initialize();
+  set->mImpl = std::move(impl);
+  return set.forget();
+}
+
 JSObject* FontFaceSet::WrapObject(JSContext* aContext,
                                   JS::Handle<JSObject*> aGivenProto) {
   return FontFaceSet_Binding::Wrap(aContext, this, aGivenProto);